fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java

```diff
@@ -648,11 +674,58 @@ private TopDocs getPageOfData(
             int perPage,
             int currentPage
     ) {
+        /**
+         * Works around the lack of a way to get at the underlying Collector in TimeLimitingCollector
+         */
+        class AccessibleTimeLimitingCollector extends TimeLimitingCollector {
+            private TopScoreDocCollector wrappedCollector;
+
+            /**
+             * @param collector the wrapped collector
+             * @param clock the clock used to implement the timeout
+             * @param ticksAllowed the amount of time allowed for the search
+             */
+            AccessibleTimeLimitingCollector(TopScoreDocCollector collector, Counter clock, long ticksAllowed) {
+                super(collector, clock, ticksAllowed);
+                wrappedCollector = collector;
+            }
+
+            /**
+             * This is the method TimeLimitingCollector was missing.
+             * @return the wrapped collector
+             */
+            public TopScoreDocCollector getWrappedCollector() {
+                return wrappedCollector;
+            }
+        }
+
+        /**
+         * This parallelizes the lucene search - each thread using a time limited collector.
+         */
+        final CollectorManager<AccessibleTimeLimitingCollector, TopDocs> manager
+                = new CollectorManager<AccessibleTimeLimitingCollector, TopDocs>() {
+
+            @Override
+            public AccessibleTimeLimitingCollector newCollector() throws IOException {
+                return new AccessibleTimeLimitingCollector(
+                        TopScoreDocCollector.create(perPage, lastEntry), Counter.newCounter(false), searchTimeout
+                );
+            }
+
+            @Override
+            public TopDocs reduce(Collection<AccessibleTimeLimitingCollector> collectors) throws IOException {
+                final TopDocs[] docs = collectors.stream()
+                        .map((collector) -> collector.getWrappedCollector().topDocs())
+                        .collect(Collectors.toList())
+                        .toArray(new TopDocs[collectors.size()]);
+
+                return TopDocs.merge(perPage, docs);
+            }
+        };
+
         lock.readLock().lock();
         try {
-            return lastEntry == null ?
-                    indexSearcher.search(query, perPage) :
-                    indexSearcher.searchAfter(lastEntry, query, perPage);
+            return indexSearcher.search(query, manager);
```

Comment:
Yeah, I looked closely at this too, and the code we have here is almost exactly the same logic that the TopN collector is doing internally. The difference is that, by using a manager, we're actually parallelizing the work across more than 1 thread and the reducing down, which we were not doing before, so things _should_ be faster (and more concretely bounded).

The only worry I have is that having more disjoint thread pools (ie. Lucene will use it's own for this parallelism) makes it harder to tune things, since you can have more contention for processors more easily, but if we see that, we can deal with it.